walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,123 @@
+
+    if(takenTimesMerged.size()>1) {
+      for (int i = 0; i < takenTimesMerged.size(); i++) {
+        if(i == takenTimesMerged.size()-1) {
+          break;
+        }
+        TimeRange currentRange = takenTimesMerged.get(i);
+        TimeRange adjRange = takenTimesMerged.get(i+1);
+        int newStart = currentRange.end();
+        int newEnd = adjRange.start();
+        validTimes.add(TimeRange.fromStartEnd(newStart,newEnd,false));
+      }
+    }
+    return validTimes;
+  }
+
+  private ArrayList<TimeRange> filterTimes(ArrayList<TimeRange> validTimes, MeetingRequest request) {
+    ArrayList<TimeRange> finalTimes = new ArrayList();
+    for(TimeRange time : validTimes) {
+      if(time.duration() >= request.getDuration()) {
+          finalTimes.add(time);
+      }
+    }
+    Collections.sort(finalTimes, TimeRange.ORDER_BY_START);
+    return finalTimes;
+  }
+  /**
+   * Parameters: a collection of events occuring that day, and a meeting request containings
+   * a time for the meeting, duration, and attendence list of manadatory and optional attenedees
+   * a time for the meeting,
+   * 
+   * Response: Based on the events happeing on the day, query will find and return time ranges
+   * in which the meeting can be hosted because no mandatory members will have a conflict.
+   * It will try to factor in optional attendees so long as they do not cause the mandatory
+   * members to not attend. 
+   */
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+    if(request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList();
+    }
+    if(events.isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    ArrayList<TimeRange> takenTimes = new ArrayList();
+    ArrayList<TimeRange> takenTimesOptional = new ArrayList();
+    for(Event event : events) {
+      if(!Collections.disjoint(event.getAttendees(),request.getAttendees())) {
+          takenTimes.add(event.getWhen());
+        }
+      if(!Collections.disjoint(event.getAttendees(),request.getOptionalAttendees())) {
+          takenTimesOptional.add(event.getWhen());
+        }
+    }
+
+    if(takenTimes.isEmpty() && takenTimesOptional.isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    if(!takenTimes.isEmpty()) {
+      ArrayList<TimeRange> takenTimesMerged = getMergedTimes(takenTimes);
+      ArrayList<TimeRange> validTimes = getBetweenTimes(takenTimesMerged);
+      ArrayList<TimeRange> finalTimes = filterTimes(validTimes, request);
+      ArrayList<TimeRange> remove = new ArrayList();
+
+      for(TimeRange optional: takenTimesOptional) {
+        for(TimeRange mandatory: finalTimes) {
+          if(mandatory.overlaps(optional)) {
+              remove.add(mandatory);
+            }
+          }
+        }
+        if(remove.size()!= finalTimes.size()) {
+          finalTimes.removeAll(remove);
+          return finalTimes;
```

Comment:
You can place `return finalTimes;` after the if/else since it happens on either branch.